src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -625,102 +610,73 @@ namespace xclhwemhal2 {
-            address_range = convert(xml_kernel_info.second.get<std::string>("<xmlattr>.range"));
-            if (mLogStream.is_open())
-              mLogStream << __func__ << " Getting the Address Range of mode : " << mode << " info from xclbin.xml" << std::endl;
-          }
+        //fetch instance name
+        instance_name = cu.get_name();
+        //iterate over arguments and populate kernelArg structure
+        for (auto arg : args)
+        {
+          std::string arg_name = arg.get_name();
+          std::string arg_port = arg.get_port();
+          uint64_t arg_offset = arg.get_offset();
+          uint64_t arg_size = arg.get_size();
+          size_t arg_index = arg.get_index();
+          KernelArg k_arg;
+          k_arg.name = props.name + ":" + arg_name;
+          k_arg.size = arg_size;
+          kernelArgInfo[arg_offset] = k_arg;
+          if (mLogStream.is_open())
+            mLogStream << __func__ << " Filling kernel Args name: " << arg_name << " index: " << arg_index << " port: " <<
+            arg_port << " info from xclbin.xml" << std::endl;
         }
-
-        if (xml_kernel_info.first == "instance")
+        if (xclemulation::config::getInstance()->isMemLogsEnabled())
         {
-          instanceName = xml_kernel_info.second.get<std::string>("<xmlattr>.name");
-          for (auto& xml_remap : xml_kernel_info.second)
-          {
-            if (xml_remap.first != "addrRemap")
-              continue;
-
-            if (mLogStream.is_open())
-              mLogStream << __func__ << " Filling kernel Instance info from xclbin.xml" << std::endl;
-
-            uint64_t base = convert(xml_remap.second.get<std::string>("<xmlattr>.base"));
-            mCuBaseAddress = base & 0xFFFFFFFF00000000;
-
-            std::string vbnv  = mDeviceInfo.mName;
-            //BAD Worharound for vck5000 need to remove once SIM_QDMA supports PCIE bar
-            if(xclemulation::config::getInstance()->getCuBaseAddrForce()!=-1) {
-              mCuBaseAddress = xclemulation::config::getInstance()->getCuBaseAddrForce();
-            } else if(mVersalPlatform) {
-              mCuBaseAddress = 0x20200000000;
-            }
-            mKernelOffsetArgsInfoMap[base] = kernelArgInfo;
-            if (xclemulation::config::getInstance()->isMemLogsEnabled())
-            {
-              std::ofstream* controlStream = new std::ofstream;
-              controlStream->open(instanceName + "_control.mem");
-              mOffsetInstanceStreamMap[base] = controlStream;
-            }
-            break;
-          }
+          std::ofstream* controlStream = new std::ofstream;
+          //trim instance_name to get actual instance name
+          size_t position = instance_name.find(":");   
+          std:: string trimmed_instance_name = instance_name.substr(position+1);
+          controlStream->open(trimmed_instance_name + "_control.mem");
+          mOffsetInstanceStreamMap[base_address] = controlStream;
```

Comment:
Alright, I see you manage the stream object here, but still it shouldn't be a floating heap object because ownership is not obvious.   If need be, then construct as shared pointer to express shared ownership, or unique_ptr to express exclusive ownership, or non heap object and std::move it on line 666.